File: lectures/10-binary search/binary-search-nhi/python/interview_qs/ceiling.py
# Given a sorted array, find the smallest element >= the target
def main():
    arr = [2, 3, 5, 9, 14, 16, 18]
    target = 15
    print(ceiling(arr, target))

# copy past BS
# return start instead of mid   
def ceiling(arr, target):
    # check if target > biggest # in array
    if target > arr[len(arr) - 1]:
        return -1

    startIndex = 0
    endIndex = len(arr) - 1

    while startIndex <= endIndex:
        # find middle index
        midIndex = round(startIndex + (endIndex - startIndex) / 2)
        print(f"middle: {midIndex}")

        if target < arr[midIndex]:
            # check left side
            endIndex = midIndex - 1
        elif target > arr[midIndex]:
            # check right side
            startIndex = midIndex + 1
        else:
            # target = mid
            return arr[midIndex]

    return arr[startIndex]
    
if __name__ == "__main__":
    main()

File: lectures/10-binary search/binary-search-nhi/binary-search-nhi/src/com/nhi/inteviewQs/Ceiling.java
package com.nhi.inteviewQs;

/**
 * Given a sorted array, find the smallest element >= the target
 */
public class Ceiling {
    public static void main(String[] args) {
        int[] arr = {2, 3, 5, 9, 14, 16, 18};
        int target = 15;
        int ans = ceiling(arr, target);
        System.out.println(ans);
    }

    /**
     * we just copy paste the BS code, and return start instead of -1
     * @param arr
     * @param target
     * @return
     */
    // return index
    // return -1 if DNE
    static int ceiling(int[] arr, int target) {
        // but what if the target > greatest # in array?
        if (target > arr[arr.length - 1]) {
            return -1;
        }
        int startIndex = 0;
        int endIndex = arr.length - 1;

        while (startIndex <= endIndex) {
            // find middle elem
            // int mid = (start + end) / 2;
            // it's possible that start + end exceeds the range for int
            int midIndex = startIndex + (endIndex - startIndex) / 2;

            if (target < arr[midIndex]) {
                endIndex = midIndex - 1; // check the left side
            } else if (target > arr[midIndex]) {
                startIndex = midIndex + 1; // check the right side
            } else {
                // ans found
                return arr[midIndex];
            }
        }
        return arr[startIndex];
    }
}
